src/index.test.ts

```diff
@@ -1,6 +1,1431 @@
 import * as ethUtil from 'ethereumjs-util';
 import * as sigUtil from '.';
 
+const encodeDataExamples = {
+  // dynamic types supported by EIP-712:
+  bytes: [10, '10', '0x10', Buffer.from('10', 'utf8')],
+  string: [
+    'Hello!',
+    '0xbBbBBBBbbBBBbbbBbbBbbbbBBbBbbbbBbBbbBBbB',
+    '0xabcd',
+    '😁',
+    10,
+  ],
+  // atomic types supported by EIP-712:
+  address: [
+    '0xbBbBBBBbbBBBbbbBbbBbbbbBBbBbbbbBbBbbBBbB',
+    '0x0',
+    10,
+    'bBbBBBBbbBBBbbbBbbBbbbbBBbBbbbbBbBbbBBbB',
+    Number.MAX_SAFE_INTEGER,
+  ],
+  bool: [true, false, 'true', 'false', 0, 1, -1, Number.MAX_SAFE_INTEGER],
+  bytes1: [
+    '0x10',
+    10,
+    0,
+    1,
+    -1,
+    Number.MAX_SAFE_INTEGER,
+    Buffer.from('10', 'utf8'),
+  ],
+  bytes32: [
+    '0x10',
+    10,
+    0,
+    1,
+    -1,
+    Number.MAX_SAFE_INTEGER,
+    Buffer.from('10', 'utf8'),
+  ],
+  int8: [0, '0', '0x0', 255, -255],
+  int256: [0, '0', '0x0', Number.MAX_SAFE_INTEGER, Number.MIN_SAFE_INTEGER],
+  uint8: [0, '0', '0x0', 255],
+  uint256: [0, '0', '0x0', Number.MAX_SAFE_INTEGER],
+  // atomic types not supported by EIP-712:
+  int: [0, '0', '0x0', Number.MAX_SAFE_INTEGER, Number.MIN_SAFE_INTEGER], // interpreted as `int256` by `ethereumjs-abi`
+  uint: [0, '0', '0x0', Number.MAX_SAFE_INTEGER], // interpreted as `uint256` by `ethereumjs-abi`
+  // `fixed` and `ufixed` types omitted because their encoding in `ethereumjs-abi` is very broken at the moment.
+  // `function` type omitted because it is not supported by `ethereumjs-abi`.
+};
+
+const encodeDataErrorExamples = {
+  address: [
+    {
+      input: '0xbBbBBBBbbBBBbbbBbbBbbbbBBbBbbbbBbBbbBBbB0',
+      errorMessage: 'Supplied uint exceeds width: 160 vs 164',
+    },
+  ],
+  int8: [{ input: '256', errorMessage: 'Supplied int exceeds width: 8 vs 9' }],
+  uint: [{ input: -1, errorMessage: 'Supplied uint is negative' }],
+  uint8: [{ input: -1, errorMessage: 'Supplied uint is negative' }],
+  uint256: [{ input: -1, errorMessage: 'Supplied uint is negative' }],
+  bytes1: [
+    { input: 'a', errorMessage: 'Cannot convert string to buffer' },
+    { input: 'test', errorMessage: 'Cannot convert string to buffer' },
+  ],
+  bytes32: [
+    { input: 'a', errorMessage: 'Cannot convert string to buffer' },
+    { input: 'test', errorMessage: 'Cannot convert string to buffer' },
+  ],
+};
+
+// Union of all types from both sets of examples
+const allExampleTypes = [
+  ...new Set(
+    Object.keys(encodeDataExamples).concat(
+      Object.keys(encodeDataErrorExamples),
+    ),
+  ),
+];
+
+describe('TypedDataUtils.encodeData', function () {
```

Comment:
Why are we not using arrow functions for the callbacks in this test suite? `mocha` hangover?

Reply:
Ah, right :sweat_smile: 
I was just following existing conventions (I think?). Probably better to use arrow functions, but maybe we can do this in one pass later on.